### PR TITLE
Integration test for node selection (PR)

### PR DESCRIPTION
**Pull request by deyanzz**
**Description**:

**Related issue(s)**:

**Fixes:** https://github.com/hashgraph/hedera-sdk-cpp/issues/167

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Comment by rwalworth** [comment]
Hey @deyanzz what's going on with this PR? It was opened a while ago but I'm not sure if it's complete/ready to review

**Comment by deyanzz** [comment]
> Hey @deyanzz what's going on with this PR? It was opened a while ago but I'm not sure if it's complete/ready to review

Yes, I was hanging around with this PR for too long. I was wondering how exactly to implement the algorithm for node selection. I was searching for a similar tests in the [Java SDK](https://github.com/hashgraph/hedera-sdk-java), but I found only one test to change the network in [ClientTest.java](https://github.com/hashgraph/hedera-sdk-java/blob/cf1c1e32c935af9fb33b65cc14c55df4902ec02a/sdk/src/test/java/com/hedera/hashgraph/sdk/ClientTest.java#L160).
Unfortunately, I wasn't able to find any kind of integration test for node selection in the [JavaScript SDK](https://github.com/hashgraph/hedera-sdk-js).

Definitely, it's better to start a discussion in this PR to think about the best algorithm for node selection.

**Comment by rwalworth** [comment]
> > Hey @deyanzz what's going on with this PR? It was opened a while ago but I'm not sure if it's complete/ready to review
> 
> Yes, I was hanging around with this PR for too long. I was wondering how exactly to implement the algorithm for node selection. I was searching for a similar tests in the [Java SDK](https://github.com/hashgraph/hedera-sdk-java), but I found only one test to change the network in [ClientTest.java](https://github.com/hashgraph/hedera-sdk-java/blob/cf1c1e32c935af9fb33b65cc14c55df4902ec02a/sdk/src/test/java/com/hedera/hashgraph/sdk/ClientTest.java#L160). Unfortunately, I wasn't able to find any kind of integration test for node selection in the [JavaScript SDK](https://github.com/hashgraph/hedera-sdk-js).
> 
> Definitely, it's better to start a discussion in this PR to think about the best algorithm for node selection.

The algorithm is already implemented in `Network::getNodeAccountIdsForExecute()`, the tests just need to make sure that correct nodes are picked. I'm thinking for now it may be best to wait to implement these tests after we are able to simulate node connectivity behavior (i.e. we're able to control multiple nodes in a local setup and are able set them to be healthy/unhealthy). Having this will allow us to test the full breadth of node selection. I'd recommend closing this PR for now and putting this task on the back burner until we have that functionality, which I believe will be a part of the FST tool that the `hedera-local-node` team is working on at the moment.

**Comment by rwalworth** [comment]
@deyanzz are you good if I close this for now?

**Comment by deyanzz** [comment]
> @deyanzz are you good if I close this for now?

Sorry for my late response. I'm preparing a commit, where will introduce a new test similar to [setNetworkWorks](https://github.com/hashgraph/hedera-sdk-java/blob/cf1c1e32c935af9fb33b65cc14c55df4902ec02a/sdk/src/test/java/com/hedera/hashgraph/sdk/ClientTest.java#L160) from the [Java SDK](https://github.com/hashgraph/hedera-sdk-java).

Also, I've done some minor refactoring for `ClientIntegrationTest`.

**Comment by deyanzz** [comment]
With commit 191d6506134a21971575ac68acf8da885625be1b I moved all of the unit tests for min & max backoff time into `ClientTest`.
Also, I'm suggesting to add a new integration test `SetNetworkIsWorkingCorrectly` in `ClientIntegrationTest`.

Created new [issue #522 ](https://github.com/hashgraph/hedera-sdk-cpp/issues/522) to track missing integration test for node selection.

File: sdk/tests/integration/ClientIntegrationTest.cc [hunk 1]
```diff
@@ -47,22 +61,14 @@ class ClientIntegrationTest : public ::testing::Test
   [[nodiscard]] inline const AccountId& getAccountId() const { return mAccountId; }
   [[nodiscard]] inline const std::string getPathToJSON() const { return mFilePath.string(); }
 
-  [[nodiscard]] inline const std::chrono::milliseconds getNegativeBackoffTime() const { return mNegativeBackoffTime; }
-  [[nodiscard]] inline const std::chrono::milliseconds getZeroBackoffTime() const { return mZeroBackoffTime; }
-  [[nodiscard]] inline const std::chrono::milliseconds getBelowMinBackoffTime() const { return mBelowMinBackoffTime; }
-  [[nodiscard]] inline const std::chrono::milliseconds getAboveMaxBackoffTime() const { return mAboveMaxBackoffTime; }
-
 private:
+  // Client mClient;
```
Comment (rwalworth) [inline sdk/tests/integration/ClientIntegrationTest.cc:65]: Should get rid of this commented-out code.

File: sdk/tests/integration/ClientIntegrationTest.cc [hunk 2]
```diff
@@ -17,27 +17,41 @@
...
 
 using json = nlohmann::json;
+using namespace std;
 using namespace Hedera;
 
-class ClientIntegrationTest : public ::testing::Test
+class ClientIntegrationTest : public BaseIntegrationTest
 {
 protected:
+  // [[nodiscard]] inline const Client& getTestClient() const { return mClient; }
```
Comment (rwalworth) [inline sdk/tests/integration/ClientIntegrationTest.cc:54]: Should get rid of this commented-out code.

File: sdk/tests/integration/ClientIntegrationTest.cc [hunk 3]
```diff
@@ -124,52 +130,36 @@ TEST_F(ClientIntegrationTest, ConnectToLocalNode)
...
+  client.setNetwork(newNetworkMap);
 
   // When / Then
-  EXPECT_NO_THROW(client.setMaxBackoff(DEFAULT_MIN_BACKOFF));
-  EXPECT_NO_THROW(client.setMaxBackoff(DEFAULT_MAX_BACKOFF));
-}
\ No newline at end of file
+  ASSERT_NO_THROW(accountBalance_6 = AccountBalanceQuery().setAccountId(accountId_6).execute(client));
+  ASSERT_NO_THROW(accountBalance_7 = AccountBalanceQuery().setAccountId(accountId_7).execute(client));
+}
```
Comment (rwalworth) [inline sdk/tests/integration/ClientIntegrationTest.cc:165]: This is fine for now but I think we should really try to enforce going forward that the integration tests do not communicate with previewnet/testnet/mainnet nodes, hence my previous comments about waiting for the multi-node setup to be complete. A user should be able to test locally with that set-up if they so desire, and seeing as how this test contains URLs for testnet consensus nodes, these tests would not pass.

**Review by rwalworth — APPROVED** [review]
LGTM, but there should be some follow-up on these once the multi-node test setup is done.

File: sdk/tests/integration/ClientIntegrationTest.cc (hunk 1 above)
Comment (deyanzz) [reply sdk/tests/integration/ClientIntegrationTest.cc]: Removed in b8259a713eabcbeb554889a5e5e9a602f4690bac

File: sdk/tests/integration/ClientIntegrationTest.cc (hunk 3 above)
Comment (deyanzz) [reply sdk/tests/integration/ClientIntegrationTest.cc:162]: I totally agree with you - using hardcoded URLs in the integration tests is a bad practice.
Integration tests should communicate only with the local node and this approach should be used in all of the Hedera's SDKs.
For now I will suggest to add a new JSON config files containing all of the consensus nodes for mainnet / previewnet / test. I will track this change in https://github.com/hashgraph/hedera-sdk-cpp/issues/525.

Comment (rwalworth) [reply sdk/tests/integration/ClientIntegrationTest.cc:162]: I don't think this is necessary as we should not be communicating with testnet nodes to begin with. We would add this JSON file ultimately to remove it when we get the multi-node setup integrated. I think we should leave it for now and then go through this again when we get the multi-node setup working.

**Pull request merged**